File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/B-Взрывоопасность/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    A: list[int] = [0] * (n + 1)
    B: list[int] = [0] * (n + 1)
    C: list[int] = [0] * (n + 1)
    A[1] = B[1] = C[1] = 1
    for i in range(2, n + 1):
        A[i] = B[i - 1] + C[i - 1]
        B[i] = A[i - 1] + B[i - 1] + C[i - 1]
        C[i] = A[i - 1] + B[i - 1] + C[i - 1]
    sys.stdout.write('{}\n'.format(A[n] + B[n] + C[n]))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/B-Взрывоопасность/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<long long> A(n + 1, 0), B(n + 1, 0), C(n + 1, 0);
    A[1] = 1;
    B[1] = 1;
    C[1] = 1;
    for (int i = 2; i <= n; ++i) {
        A[i] = B[i - 1] + C[i - 1];
        B[i] = A[i - 1] + B[i - 1] + C[i - 1];
        C[i] = A[i - 1] + B[i - 1] + C[i - 1];
    }
    std::printf("%lld\n", A[n] + B[n] + C[n]);
}


int main() {
    solution();
}